### PR TITLE
Add niriss soss datamodel

### PR DESCRIPTION
**Pull request by sosey**
Closes #

Resolves JP-

**Description**

This PR comes from a discussion with @eslavich about reference file validation and adds a datamodel for the niriss SOSS wavelength tables, the specwcs reference file...it likely still needs work, tests, a doc update and should be reviewed by @nden  or someone familiar with niriss soss mode before acceptance.  There's an example new reference file as well on CS. 

Checklist
- [ ] Tests
- [ ] Documentation
- [ ] Change log
- [x] Milestone
- [x] Label(s)

### CODE REVIEW

File: jwst/assign_wcs/niriss.py [hunk 1]
```diff
@@ -108,12 +110,20 @@ def niriss_soss(input_model, reference_files):
...
     -----
-    It includes tWO coordinate frames -
+    It includes two coordinate frames -
     "detector" and "world".
 
-    It uses the "specwcs" reference file.
+    It uses the "specwcs" reference file type.
     """
 
+    # The input is the grism image
```
Comment (hbushouse) [inline jwst/assign_wcs/niriss.py:119]: Which kind of NIRISS "grism" image are we talking about here? The GR700XD grism that's used for SOSS mode, or one of the GR150R or GR150C grisms that's used for WFSS mode?

File: jwst/assign_wcs/niriss.py [hunk 2]
```diff
@@ -108,12 +110,20 @@ def niriss_soss(input_model, reference_files):
...
+    if not isinstance(input_model, ImageModel):
```
Comment (hbushouse) [inline jwst/assign_wcs/niriss.py:120]: SOSS mode data are usually taken as a time-series observation, which are always processed on a per-integration basis, and hence the input to the spec2 and spec3 pipeline levels is expected to be a `CubeModel` (3-D stack of integrations), not an `ImageModel`. Unfortunately, it's not mandatory that SOSS uses the time-series approach, so the input could be either `ImageModel` (2D) or `CubeModel` (3D). You never know.

File: jwst/assign_wcs/niriss.py [hunk 3]
```diff
@@ -133,18 +143,12 @@ def niriss_soss(input_model, reference_files):
...
-            bytesio_or_path = reference_files['specwcs']
-        with asdf.open(bytesio_or_path) as af:
-            wl1 = af.tree[1].copy()
-            wl2 = af.tree[2].copy()
-            wl3 = af.tree[3].copy()
-    except Exception as e:
+        with NirissSOSSTableModel(reference_files['specwcs']) as f:
+            wl1 = f.models[0]  # order 1
+            wl2 = f.models[1]  # order 2
+            wl3 = f.models[2]  # order 3
```
Comment (hbushouse) [inline jwst/assign_wcs/niriss.py:149]: Is it guaranteed that data for all 3 orders will always be present in the ref file?

File: jwst/datamodels/__init__.py [hunk 4]
```diff
@@ -79,7 +79,7 @@
                              RegionsModel, WavelengthrangeModel, CameraModel, CollimatorModel, OTEModel,
                              FOREModel, FPAModel, IFUPostModel, IFUFOREModel, IFUSlicerModel, MSAModel,
                              FilteroffsetModel, DisperserModel, NIRCAMGrismModel, NIRISSGrismModel,
-                             WaveCorrModel)
+                             NirissSOSSTableModel,WaveCorrModel)
```
Comment (hbushouse) [inline jwst/datamodels/__init__.py:82]: I'm guessing flake8 won't like the missing white space after the comma here

File: jwst/assign_wcs/niriss.py [hunk 5]
```diff
@@ -133,18 +143,12 @@ def niriss_soss(input_model, reference_files):
...
+
+    except IOError as e:
         raise IOError(f"Error reading wavelength correction from {reference_files['specwcs']}") from e
```
Comment (hbushouse) [inline jwst/assign_wcs/niriss.py:152]: Minor nit and not part of the PR changes, but is it technically correct to refer to the data in the ref file as a wavelength "correction"? I assumed it was just data that defined the wavelength scale to begin with, as opposed to adding some kind of correction to it.

File: jwst/datamodels/wcs_ref_models.py [hunk 6]
```diff
@@ -341,6 +341,56 @@ def to_fits(self):
...
+        self.meta.instrument.detector = "NIS"
+        self.meta.exposure.type = "NIS_SOSS"
+        self.meta.reftype = self.reftype
+
+    def validate(self):
+        super(NirissSOSSTableModel, self).validate()
+        try:
+            assert isinstance(self.meta.input_units, (str, u.NamedUnit))
+            assert isinstance(self.meta.output_units, (str, u.NamedUnit))
+            assert self.meta.instrument.name == "NIRISS"
```
Comment (hbushouse) [inline jwst/datamodels/wcs_ref_models.py:377]: Given that all of these values are hardwired just 10 lines above, is it even possible for the asserts to ever fail and hence are they really useful?

**Review by hbushouse — COMMENTED** [review]
The new model itself looks OK to me, but lots of comments/questions elsewhere.

**Comment by melanieclarke** [comment]
@tapastro - this PR looks like it is probably no longer needed, or else would need significant work to bring it up to date.  Is it okay to close, or does this need a ticket filed to follow up?

**Comment by hbushouse** [comment]
If you do end up implementing this, be sure to pay attention to the extremely useful and wise comments left by a reviewer about 3 years ago!  ;-)

**Comment by sosey** [comment]
I have zero memory of any of this 😆

**Comment by tapastro** [comment]
This PR has an aura about it - I don't recall seeing this ping either...

Closing the PR. It had a good run. No longer applicable, three years later.
